ngcplogger.go

```diff
@@ -345,6 +352,40 @@ func (l *nGCPLogger) extractMsgFromPayload(m map[string]any) {
 	}
 }
 
+func (l *nGCPLogger) extractGcpFromPayload(m map[string]any, entry *logging.Entry) {
+
+	if l.extractGcp {
+		if val, exists := m["logging.googleapis.com/sourceLocation"]; exists {
+			v := val.(map[string]any)
+			entry.SourceLocation = &loggingpb.LogEntrySourceLocation{
+				File:     v["file"].(string),
+				Line:     int64(v["line"].(float64)),
+				Function: v["function"].(string),
+			}
+			delete(m, "logging.googleapis.com/sourceLocation")
+		}
+		if val, exists := m["logging.googleapis.com/trace"]; exists {
+			entry.Trace = val.(string)
+			delete(m, "logging.googleapis.com/trace")
+		}
+		if val, exists := m["logging.googleapis.com/spanId"]; exists {
+			entry.SpanID = val.(string)
+			delete(m, "logging.googleapis.com/spanId")
+		}
+		if val, exists := m["logging.googleapis.com/trace_sampled"]; exists {
+			entry.TraceSampled = val.(bool)
+			delete(m, "logging.googleapis.com/trace_sampled")
+		}
+		if val, exists := m["logging.googleapis.com/labels"]; exists {
+			v := val.(map[string]any)
+			for k, v := range v {
+				entry.Labels[k] = v.(string)
+			}
+			delete(m, "logging.googleapis.com/labels")
+		}
+	}
```

Comment:
I think it would be better if the casting from any types was done while receiving the second optional param from the cast operation and checking that it went well. This way, we avoid a panic in case the underlying type does not match the type you want to cast to, if something changed in the future. Otherwise we risk every log crashing the driver if any field changed type. It should be fairly simple to change, you should be able to make a generic function for this.

Reply:
OK, I left these casts explicitly open so it would panic and I see any deficiencies right away. So doing it conditionally would also require logging a mismatch somewhere, silently ignoring stuff would probably not a good idea. As I do seldomly look into the logs of the docker daemon anymore, I would think appending something to the log record being processed would show better that something needs too be done?

Reply:
Yeah we definitely don't want to ignore those errors. I think tracing the error in the driver slogger should be fine, if you were expecting to see it and you don't, I assume the user will check the logging driver logs. We could also consider appending a field to the log with the error that occurred as you suggest, or maybe even better, generating a separate log entry when processing a log fails, so we can assign it a different severity level and the user can assign the severity they want to those messages. I would make this optional and maybe not enable it by default, to avoid polluting the users logs (storage does cost after all, and maybe you don't want to be suddenly spammed by all these errors).

But I think this is out of scope for the changes you are proposing here. Maybe it would be easier to just return an error for now, which will be logged by the internal driver logger, and we can open an issue to work on improving the error handling in general across the driver separately, so we can find a good solution for reporting errors and we don't block this issue. What do you think?

Reply:
I have updated the pull request so it does not panic any longer but does log the problem for finding and fixing any issues. What do you think?

Reply:
Looks good, we can leave it that way for now and open an issue to discuss a more definite solution.